File: src/jupyter_matlab_labextension/src/plugins/actions/matlabStatusAction.ts
// Copyright 2026 The MathWorks, Inc.

import { PromiseDelegate, ReadonlyJSONValue } from '@lumino/coreutils';

import { ICommunicationChannel } from '../matlabCommunication';
import { BaseAction } from './baseAction';
import { ActionTypes } from './actionTypes';
import { isCommValid } from './actionUtils';

export type MATLABStatus = {
  isMatlabLicensed: boolean;
  matlabStatus: string;
  matlabProxyHasError: boolean;
  licensingMode: string;
  matlabVersion: string;
  matlabRootPath: string;
};

export class MatlabStatusAction extends BaseAction {
    private static status: MATLABStatus;
    private static blockingPromise: PromiseDelegate<ReadonlyJSONValue> | null;

    constructor (blocking: boolean) {
        super();
        this.blocking = blocking;
    }

    public static getStatus (): MATLABStatus {
        return MatlabStatusAction.status;
    }

    public getActionName (): string {
        return ActionTypes.MATLAB_STATUS;
    }

    public async execute (data: any, comm: ICommunicationChannel): Promise<void> {
        if (!MatlabStatusAction.blockingPromise) {
            MatlabStatusAction.blockingPromise =
        new PromiseDelegate<ReadonlyJSONValue>();
        }

        console.debug('MatlabStatusAction execute called with data ', data);
        this.fetchMatlabStatus(comm);

        if (this.blocking) {
            console.debug(
                'This is a blocking action will wait for promise to resolve...'
            );
            await MatlabStatusAction.blockingPromise.promise;
        }
        console.debug('MatlabStatusAction execute completed');
    }

    public onMsg (data: any, _: ICommunicationChannel): void {
        console.debug('Received data for matlab_status action from kernel ', data);
        MatlabStatusAction.status = data.data as MATLABStatus;

        if (MatlabStatusAction.blockingPromise) {
            console.debug('Resolved blocking promise...');
            MatlabStatusAction.blockingPromise.resolve(null);
            MatlabStatusAction.blockingPromise = null;
        }

        console.debug('MatlabStatusAction onMsg completed');
    }

    private fetchMatlabStatus (comm: ICommunicationChannel): any {
        if (!isCommValid(comm)) {
            console.error('Communication channel is not available');
            return false;
        }

        comm.send({
            action: 'matlab_status',
            data: {}
        });
    }
}

File: src/jupyter_matlab_kernel/kernels/labextension_comm/actions/matlab_status_action.py
# Copyright 2025 The MathWorks, Inc.

from . import ActionCommand, ActionTypes


class MatlabStatusAction(ActionCommand):
    def __init__(self, kernel):
        self.kernel = kernel
        self.log = kernel.log

    def get_code(self):
        pass

    async def execute(self, comm, data):
        try:
            self.log.debug("Fetching MATLAB proxy status...")
            (
                is_matlab_licensed,
                matlab_status,
                _,
            ) = await self.kernel.mwi_comm_helper.fetch_matlab_proxy_status()

            self.log.debug("MatlabStatus action successful")
            comm.send(
                {
                    "action": ActionTypes.MATLAB_STATUS.value,
                    "matlabStatus": {
                        "isLicensed": is_matlab_licensed,
                        "status": matlab_status,
                    },
                    "error": None,
                }
            )
        except Exception as err:
            self.log.error(f"Error fetching MATLAB proxy status: {err}")
            comm.send(
                {
                    "action": ActionTypes.MATLAB_STATUS.value,
                    "matlabStatus": {},
                    "error": str(err),
                }
            )
